### PR TITLE
Refactor skin construction

### PR DESCRIPTION
**Pull request by peppy**
## Preface

As I continue work on the skin editor, I added support for drag-and-drop sprite importing. While doing so, I noticed that `DefaultSkin`s which have been customised by the user didn't have access to `Sample`s or `Texture`s because that logic is all local to `LegacySkin`.

This PR attempts to bring some sanity to how skins are constructed, but also makes fixing the above quite trivial (have tested that this [works](https://github.com/ppy/osu/compare/master...peppy:skin-editor-sprites?expand=1) with a few more [hacky test lines](https://github.com/ppy/osu/commit/7b41847c2201fea4e9729deb64f7f1a0dfd90e87) of change).

## Testing

I've been relying strictly on unit tests for testing this. Probaly need some more manual testing, especially around configuration fallbacks. Mainly wanted to get this out to ensure from a code-quality perspective everyone is happy, since this will be the start of more changes in a similar direction.

## Changes

Individual changes are split out into commits, but the important ones are listed below:

### 5c4a74378d Remove `Textures` and `Samples` initialisation to `Skin`

The main goal of this pull request.

### 078288a616 Make "skin.ini" the default skin filename and remove redundant parameters

The majority of usages were specifying `skin.ini` manually, so I believe this makes sense.

### 7a1909bf97 Change parameter order of `LegacySkin` to put `IStorageResourceProvider` first

The optional resource store should not be before the required* resource provider.

* note that the resource provider isn't strictly required because of some tests that don't make it. we probably want to clean that up in the future.

### d0ea1739b4 Remove skin configuration stream logic as it is no longer required

Because the base `Skin` implementation is now able to do realm-based lookups (or in the case of fallback, using the provided `IResourceStore`), providing a stream down the whole stack is not required. A filename is enough. Similar reasoning for 9d3c6ade62 and 6c405f1dee.

### 00aea9bef0 Only use legacy resources lookup for protected (aka default) skin

This is still a bit haphazard, and can likely be improved further in follow-up work (maybe using a `virtual IsDefault` or something?). There's an inline comment further explaining what is going on.

### e1236e07ad Fix extensions not being specified in time for realm file caching

I have follow-up work which would potentially delay the caching operation to the first lookup, or as a separate call but that requires some more restructuring (the realm store needs to receive an `ILive` otherwise any access outside its `ctor` is incorrect).

### CODE REVIEW

**Review by bdach — COMMENTED** [review]
Not going to lie, reviewing this was pretty painful. Initially opened diff, couldn't make much out of it, then switched to per-commit, which was still pretty bad (as some changes from earlier commits were reverted later down the line) but somehow got through it and then opened diff again, which was more palatable the second go-around. But I imagine this likely was even worse to write.

Notes from a brief read-through. There are some issues here but none of it seems major, I don't find any changes here particularly offensive.

**Comment by bdach** [comment]
I haven't tested this much yet, but there is a weird issue wherein if you set default legacy skin and open skin editor, all HUD elements will change back to lazer default skin somehow:

https://user-images.githubusercontent.com/20418176/159805128-0da7a85b-ced4-4376-9a5a-2c058f28cd5f.mp4

This isn't the case in `master` so I'm guessing something here must have broken it.

File: osu.Game/Skinning/Skin.cs [hunk 1]
```diff
@@ -41,16 +52,32 @@ public abstract class Skin : IDisposable, ISkin
...
+        /// <param name="configurationFilename">An optional filename to read the skin configuration from. If not provided, the configuration will be retrieved from the storage using "skin.ini".</param>
+        protected Skin(SkinInfo skin, [CanBeNull] IStorageResourceProvider resources, [CanBeNull] IResourceStore<byte[]> storage = null, [CanBeNull] string configurationFilename = @"skin.ini")
         {
-            SkinInfo = resources?.RealmAccess != null
-                ? skin.ToLive(resources.RealmAccess)
-                // This path should only be used in some tests.
-                : skin.ToLiveUnmanaged();
+            if (resources != null)
+            {
+                SkinInfo = resources.RealmAccess != null
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Skinning/Skin.cs:66]: I moved this in the follow up PR already for safety, but apparently doesn’t cause test failures as you noticed.

File: osu.Game/Skinning/LegacyBeatmapSkin.cs [hunk 2]
```diff
@@ -20,13 +21,27 @@ public class LegacyBeatmapSkin : LegacySkin
...
             Configuration.AllowDefaultComboColoursFallback = false;
         }
 
+        private static IResourceStore<byte[]> createRealmBackedStore(BeatmapInfo beatmapInfo, [CanBeNull] IStorageResourceProvider resources)
+        {
+            if (resources == null)
+                // should only ever be used in tests.
+                return new ResourceStore<byte[]>();
+
+            return new RealmBackedResourceStore(beatmapInfo.BeatmapSet, resources.Files, new[] { @"ogg" });
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Skinning/LegacyBeatmapSkin.cs:42]: The main issue is that `Skin` takes a `SkinInfo` and this is resolving from a `BeatmapSetInfo`.

I was also hoping to move it but haven’t found a path there just yet.

**Comment by peppy** [comment]
I've found the cause and fixed the above issue, but found some others (changing skin while in mania between default and classic doesn't correctly work, possibly due to the fallback lookup:

https://github.com/ppy/osu/blob/994fb966b6ba27baf7c178a7483ccf353db46fc3/osu.Game/Skinning/SkinManager.cs#L205-L217

File: osu.Game/Skinning/Skin.cs [hunk 3]
```diff
@@ -41,16 +52,32 @@ public abstract class Skin : IDisposable, ISkin
 
         public abstract IBindable<TValue> GetConfig<TLookup, TValue>(TLookup lookup);
 
-        protected Skin(SkinInfo skin, IStorageResourceProvider resources, [CanBeNull] Stream configurationStream = null)
+        /// <summary>
+        /// Construct a new skin.
+        /// </summary>
+        /// <param name="skin">The skin's metadata. Usually a live realm object.</param>
+        /// <param name="resources">Access to game-wide resources.</param>
+        /// <param name="storage">An optional store which will be used for looking up skin resources. If null, one will be created from realm <see cref="IHasRealmFiles"/> pattern.</param>
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Skinning/Skin.cs:60]: Was trying to say that the `SkinInfo`'s files will be used. Originally I tried to abstract this out further to being a realm model (to try and address this issue you [brought up here](https://github.com/ppy/osu/pull/17416#discussion_r833753917)) but can probably reword for now to be more explicit.

**Comment by peppy** [comment]
Above issues are resolved, should be ready for some further testing. I'd recommend we leave this open at least a day or two more to allow people to manually test against edge cases. We have pretty good coverage for skin lookup tests but I'm still not 100% confident.

File: osu.Game/Skinning/LegacySkin.cs [hunk 4]
```diff
@@ -385,26 +356,15 @@ public override Drawable GetDrawableComponent(ISkinComponent component)
...
-                                    }
+                                Children = new Drawable[]
+                                {
+                                    new LegacyComboCounter(),
+                                    new LegacyScoreCounter(),
+                                    new LegacyAccuracyCounter(),
+                                    new LegacyHealthDisplay(),
+                                    new SongProgress(),
+                                    new BarHitErrorMeter(),
+                                }
```
Comment (smoogipoo) [inline osu.Game/Skinning/LegacySkin.cs:367]: What's with these changes? Is it because the fallback is now being done?

Comment (peppy) [reply osu.Game/Skinning/LegacySkin.cs:367]: Yep, check commit message on https://github.com/ppy/osu/pull/17416/commits/a4d17a915f96c505e518dd261f9b4d1d5d967b7f for in depth explanation (wasn't in opening post because it was found in @bdach  review pass)

**Review by smoogipoo — APPROVED** [review]
I've done some pretty rudimentary testing with this and it seems fine.
